text/0000-single-formatter-for-all-contexts.md

```diff
@@ -0,0 +1,68 @@
+# Single Formatters for all Contexts
+
+**Status:** `proposed`
+
+Defining a single formatter which (de)serializes all contexts, rather than separate ones.
+
+## Motivation
+
+The existing propagators specification specifies that formatters should exist for each type of context (today this is SpanContext and DistributedContext). Requiring SDK authors to have to author one formatter per context type can result in error-prone and backwards-incompatible consumption and usage.
+
+The error-prone aspect comes from the configuration of said formatters. Let's say that a user is setting up their propagators by hand: in order to do so they will need to configure both their SpanContext propagators and their DistributedContext propagators. If a user wants to use a SaaS product that requires both a custom SpanContext propagator as well as a DistributedContext propagator, they will have to know to attach both. If they forget to do so, they will end up with missing context that is hard to debug (since the user of a SaaS product would probably have little internal knowledge about how propagation should occur).
+
+The backwards-incompatible changes come from the need to extract or inject context into new propagators. If a SaaS vendor or open source protocol previously had no need for on of the two propagators, but started to need it, they would need to notify all consumers that, as of this new version, the consumer must modify their code to add the additional formatter for the new context they require. This will probably come up often in SaaS offerings that previously only needed to propagate data in SpanContext, but then added their own fields that will then require custom DistributedContext formatters.
+
+## Explanation
+
+The new proposed API schema looks like:
+
+    formatters (namespace)
+        UnifiedContext.class
+        BinaryAPI.class
+        HTTPTextFormat.class
+
+With the following APIs:
+
+### UnifiedContext
```

Comment:
This isn't great from a terminology standpoint, but I think there's a distinction between:

1. a composed context that is there to help formatter authors return back only a single object
2. a context object that is designed to actually handle storage of the current context.

The UnifiedContext is not intended to be a long-lived container for all types of contex: as you mentioned there is a more general context component for that.

Ths UnifiedContext is strictly for formatters. A propagator would then take the unifiedcontext, retrieve the various context objects, then put them into the higher-level "Context" object.

The reason for the separation was to simplify and standardize the way that context is populated. For example, DistributedContext is the same throughout the process, while SpanContext can be switched out with a new span. As such, the way one may insert that value into the context may change.

I understand the desire to not attach the general context objects to the wire propagators. but something does need to eventually handle taking stuff from wire propagators and adding them to the general context.

Reply:
My preference would be to avoid an intermedia "unified context" object. The propagators themselves are still different depending on which type of context they are handling, e.g. a span context propagator really has very little in common with baggage propagator. So at a matter of the API they can both accept the underlying Context object and read the specific type of context they need.

I can see how there are some perf benefits that can be achieved on extracting contexts from the wire and combining them into the in-memory Context object if the latter supports some sort of bulk write mode (which, for example, Go context does not support). But that still means that propagators need access to such bulk write builder, not that they need to deal with an extra "unified context" container, which requires memory allocations.

Reply:
I don't think that the span context and distributed context propagator are that separate. In terms of the functionality that they provide, yes, but I don't believe this is the case in terms of vendor integrations.

For example, let's say I own a telemetry SaaS company (name it MyTel), and I want to provide you a simple OTel integration. Let's say that MyTel has it's own specification for spans, and wants to also propagate it's own token app_key. Let's say that for MyTel's product to work properly, both must exist and be propagated.

As a vendor, I want to make it really easy to integrate MyTel's telemetry into an OTEL-based SDK. If the propagators were combined, I would just have to instruct the consumer to load my integration in as a propagator. 

If they were separate, I would need to tell my users that they need to add both of those propagators. In addition, if one is not added correctly, then the user would have a very confusing issue on their hands where some of the data would be propagating correctly, and one would not.

My other reasoning is the fact that DistributedContext was designed to support CorrelationContext, which itself was designed to be an addition to the trace-context to have a more official way to supporting 3rd party labels. I think this implies that it will be very common for vendors to need to both integrate propagators with SpanContext and DistributedContext.

But regardless I think your proposal is calling to just eliminate the UnifiedContext entirely and instead use the existing Context object, which I think is fine. It does require the formatter implementer to know the getter / setter methods that are appropriate for each context, but it fulfills my primary concern.

Reply:
The other thing that concerns me a little bit about context directly is the fact that propagators are also expected to honor the EntryPropgationFilter of values to propagate, which requires all formatter authors to be aware of that implementation detail:

https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/api-distributedcontext.md#distributedcontext-propagation

But that could be a pretty easy to uphold convention, or logic added to the DistributedContext object itself.

Reply:
For MyTel to register one or all codecs it just needs to provide the user with a single registration function that will register all of them. Inside they can still be separate.

Don't know about the filters, it seems to have been added to the spec before the whole context framework is designed. We might change that.

Reply:
I think the single registration function results in an inflexible framework. In otel-python we are having discussions around the ability to support multiple different configurations and propagator chains. I think if a vendor provides effectively a magic function that doesn't better expose the correct primitives, it will make it very difficult for consumers with more complex use cases.

But this is probably just a matter of preference. I've personally been bitten by "magic" functions that do everything for you: I'd much rather prefer concise primitives.

Reply:
You're now saying the exact opposite to what you said in your MyTel example where you did want a single registration. A single catch-all function does not mean that the vendor cannot also provide fine grained individual codecs.

Reply:
Hey Yuri, I'm sorry if I'm not being clear. My proposal is for unifying the formatters such that a single formatter can act on all contexts. I was arguing in the case of MyTel that having multiple formatters can be error-prone, and we should minimize the number of boilerplate integrations that are needed to integrate with a specific SaaS provider.

Maybe my example is a poor one. I just wanted to illustrate the error-prone nature of having propagators that only work with a specific context. I believe it will be a common case to need to extract and inject from multiple contexts for integrations with SaaS providers or other backends to work as expected.

Reply:
I agree with the single formatter notion, because that's what the user code would be interacting with, which does not need to know how many actual formatters and contexts are configured behind the scene. But I am still not seeing the need for unified context because the default Context object is already acting as a unified container.